### PR TITLE
feat(training): harden local training pipeline with full test coverage and archetype integration

### PR DESCRIPTION
**Pull request by revlentless**
## Summary

- **Production-ready local training pipeline**: Complete infrastructure for running Babylon RL training locally with Atropos GRPO, including service management, checkpointing, W&B logging, and LR scheduling
- **Comprehensive test coverage**: Unit tests, JSON mode integration tests, DB integration tests, and E2E pipeline smoke tests with 4-tier testing approach
- **Archetype-aware trajectory recording**: NPCs are now assigned archetypes (trader, influencer, analyst, degen) during trajectory recording for downstream reward scoring

## Type

- [x] Feature
- [ ] Bug fix
- [ ] Refactor / cleanup (no behavior change)
- [ ] Performance
- [x] Infra / ops
- [x] Docs
- [ ] Contracts / on-chain
- [ ] Other: …

## Context / Links

- Builds on: `feat/archetype-scoring-rubrics` (archetype-specific reward weights)
- Related: Atropos GRPO training, Babylon RLAIF environment
- Docs: `packages/training/docs/TESTING.md`

## Scope (keep it focused)

- [x] This PR is focused on a single change/theme (not a catch‑all)
- [x] Drive‑by refactors are excluded or split into a separate PR
- [x] Non‑goals / follow‑ups are listed below (with links)

**Non-goals / Follow-ups:**
- Cloud GPU training deployment (separate infrastructure PR)
- Production database migration fixes (orphan migration files exist but are out of scope)
- Tinker integration for cloud-based inference

**Areas touched**
- [ ] Web UI (`apps/web`)
- [ ] Web API routes / SSE / A2A (`apps/web`)
- [ ] CLI (`apps/cli`)
- [ ] Docs site (`apps/docs`) / vendor docs (`docs/vendors/*`)
- [x] Domain / game engine (`packages/engine`, `packages/core/*`)
- [ ] Agents / runtime / A2A / MCP (`packages/agents`, `packages/a2a`, `packages/mcp`)
- [ ] API infra (`packages/api`)
- [ ] DB (`packages/db`)
- [ ] Contracts / on-chain (`packages/contracts`)
- [ ] Shared types/utils (`packages/shared`)
- [x] Tests (`packages/testing`)
- [x] Other: `packages/training` (Python training pipeline)

### Core Training Infrastructure

| File | Change |
|------|--------|
| `service_manager.py` | **NEW** - Manages Atropos API + vLLM process lifecycle with health checks, graceful shutdown, port management |
| `atropos_trainer.py` | **ENHANCED** - W&B integration, LR scheduling (constant/linear/cosine), checkpoint resume, `--skip-vllm` flag, `--judge-model` option |
| `run_training.py` | **REFACTORED** - `TrainingOrchestrator` class, proper subprocess env passthrough, validation modes, dry-run support |
| `babylon_env.py` | **ENHANCED** - Direct vLLM API calls (bypasses Atropos managed_server), deterministic judge scoring, `group_size=2` for GRPO |

### Archetype Integration

| File | Change |
|------|--------|
| `derive-archetype.ts` | **NEW** - NPC-to-archetype mapping based on personality, role, reliability |
| `trajectory-market-engine.ts` | **ENHANCED** - `archetypeResolver` callback integration |
| `generate-training-data.ts` | **ENHANCED** - Configurable `--hours` flag, archetype logging |
| `GameWorld.ts` | **ENHANCED** - Expose NPCs for archetype resolution |

### Testing Infrastructure

| File | Change |
|------|--------|
| `TESTING.md` | **NEW** - 500+ line comprehensive 4-tier testing guide |
| `docker-compose.test.yml` | **NEW** - PostgreSQL + Redis containers for integration tests |
| `test_db_integration.py` | **NEW** - 650 lines, full DB operations testing |
| `test_json_mode_integration.py` | **NEW** - 550 lines, JSON trajectory loading tests |
| `test_full_pipeline.py` | **NEW** - 640 lines, E2E smoke tests |
| `test_service_manager.py` | **NEW** - 585 lines, process lifecycle tests |
| `test_training_orchestrator.py` | **NEW** - 497 lines, orchestrator tests |
| `test_lr_scheduler.py` | **NEW** - 576 lines, LR scheduling tests |
| `ci-unit-integration.yml` | **ENHANCED** - Python test jobs added to CI |

### Utilities

| File | Change |
|------|--------|
| `schemas.py` | **NEW** - Pydantic models for trajectory validation |
| `error_recovery.py` | **NEW** - Graceful degradation utilities |
| `import_json_trajectories.py` | **NEW** - JSON → PostgreSQL trajectory import script |

## Review guide

**Start here**
- `packages/training/docs/TESTING.md` - Explains the 4-tier testing approach
- `packages/training/python/src/training/service_manager.py` - Core infrastructure
- `packages/training/src/archetypes/derive-archetype.ts` - Archetype derivation logic

**Risk**
- [ ] Low
- [x] Medium
- [ ] High

**Notes for reviewers**
- The `babylon_env.py` now calls vLLM directly via `aiohttp` instead of using Atropos's `ServerManager.managed_server()` - this was necessary because the managed_server context was hanging
- `group_size=2` is the minimum for GRPO (comparing 2 trajectories at a time)
- Python tests require `pip install -e .` in editable mode to resolve imports correctly

## Test plan

**Commands run**
- [x] `bun run check` (Biome format)
- [x] `bun run typecheck`
- [x] `bun run lint`
- [x] `bun run build`
- [x] `bun run test` (unit + integration)
- [ ] `bun run test:e2e` (if critical flows changed)
- [ ] `bun run contracts:test` (if contracts changed)
- [ ] `bun run db:check` (if DB schema/queries changed)

**Python tests run**
```bash
# Tier 1: Unit tests (no infrastructure)
cd packages/training/python
pytest tests/ -v --ignore=tests/integration/ --ignore=tests/e2e/

# Tier 2: JSON mode integration
pytest tests/integration/test_json_mode_integration.py -v

# Tier 3: DB integration (requires docker-compose.test.yml)
docker compose -f ../docker-compose.test.yml up -d
DATABASE_URL=postgresql://babylon_test:test_password@localhost:5434/babylon_test \
  pytest tests/integration/test_db_integration.py -v

# Tier 4: Full stack (GPU required)
python scripts/run_training.py --model Qwen/Qwen2.5-1.5B-Instruct --steps 1 --no-wandb
```

**Manual verification**
1. Generated training data with `bun run packages/engine/examples/generate-training-data.ts --causal --days 1`
2. Imported trajectories to test DB with `import_json_trajectories.py`
3. Ran full training step with vLLM inference

## Ops / Migration / Deployment

- [x] No deploy impact
- [ ] Requires env var updates (listed below + `.env.example` updated)
- [ ] Requires DB migration (`bun run db:migrate`) / backfill / seed
- [ ] Changes cron schedule or endpoints (`vercel.json`, `CRON_SECRET`, etc.)
- [ ] Rollout behind a flag / gradual rollout

**Notes:** This PR adds local development tooling only. No production deployment changes.

## Breaking changes

- [x] None
- [ ] Yes (describe + migration guide below)

## Security / privacy

- [x] No security impact
- [ ] Needs extra review (describe)

<details>
<summary>Area-specific notes (expand if relevant)</summary>

### Python dependencies

New dependencies in `requirements.txt`:
- `aiohttp` - For direct vLLM API calls
- `psycopg2-binary` - For PostgreSQL integration tests

### Database

The integration tests use a separate test database (`babylon_test` on port 5434) isolated from development/production. Schema is applied via `db:push` (not migrations).

### Testing tiers

| Tier | Infrastructure | Time | Coverage |
|------|----------------|------|----------|
| 1 | None | ~30s | Pure logic, mocked I/O |
| 2 | Filesystem | ~1min | JSON trajectory loading |
| 3 | Docker PostgreSQL | ~2min | Full DB operations |
| 4 | Docker + GPU | ~10min+ | End-to-end training |

</details>

## Screenshots / recordings (UI)

N/A - No UI changes

## Checklist (author)
- [x] Self-review done (diff + critical paths)
- [x] Base branch is correct (`feat/archetype-scoring-rubrics`)
- [x] Handlers remain thin and portable (validate → service → map errors)
- [x] Domain logic stays in packages (no Next/React/Elysia coupling in core)
- [ ] `.env.example` updated (if env changed) and variables documented above
- [x] Docs updated (and `bun run docs:generate` if needed)
- [x] Tests added/updated for behavior changes (or rationale provided)
- [x] Dependency changes are intentional (`bun.lock` updated)
- [ ] Code owners requested (auto via CODEOWNERS or manual)

---

## Stats

```
29 files changed, 8,312 insertions(+), 316 deletions(-)
```

**New files:** 18
**Modified files:** 11

**Test coverage added:**
- 293+ unit/integration test cases
- 4-tier testing infrastructure
- CI pipeline integration

## Summary by CodeRabbit

* **New Features**
  * Archetype-aware training & scoring, NPC archetype resolution, per-archetype rewards, and archetype rubrics/versioning.
  * GRPO training pipeline with LR schedulers (constant/linear/cosine), W&B integration, service orchestration, improved CLI, and local service manager.
  * Trajectory import tool, DB-backed data support, JSON import/validation, and vLLM/Atropos inference integration.

* **Documentation**
  * Expanded training README and a new detailed testing guide.

* **Tests**
  * Large additions: unit, integration, and end-to-end Python tests for scoring, schemas, services, and orchestrator.

* **Chores**
  * CI: new Python training test job; updated .gitignore to ignore Python venv.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<h3>Greptile Summary</h3>

- Implements production-ready local training pipeline with Atropos GRPO, including service lifecycle management, W&B logging, learning rate scheduling, and checkpoint resumption capabilities
- Adds comprehensive 4-tier testing infrastructure spanning unit tests, JSON mode integration tests, Docker-based database integration tests, and E2E pipeline validation
- Integrates archetype-aware trajectory recording where NPCs are assigned behavioral archetypes (trader, influencer, analyst, degen) during simulation for downstream reward scoring

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| `packages/training/python/src/training/babylon_env.py` | Refactored RLAIF environment with direct vLLM API calls replacing Atropos managed_server, archetype-aware scoring, and deterministic reward calculation |
| `packages/training/python/src/training/atropos_trainer.py` | Enhanced GRPO trainer with W&B integration, LR scheduling, checkpoint resume, and robust process management |
| `packages/training/python/src/training/service_manager.py` | New service orchestration system managing Atropos API and vLLM process lifecycles with health checks and graceful shutdown |
| `packages/training/src/archetypes/derive-archetype.ts` | New archetype derivation logic mapping NPC characteristics to behavioral types for training pipeline integration |
| `packages/training/python/src/training/error_recovery.py` | New comprehensive error handling and graceful degradation system for production training reliability |

<h3>Confidence score: 3/5</h3>

- This PR has significant architectural complexity with critical infrastructure changes in the training pipeline that could impact training stability
- Score reflects complex integration between TypeScript simulation engine and Python training systems, direct vLLM API calls replacing managed services, and extensive new archetype scoring logic that needs careful validation
- Pay close attention to `babylon_env.py` for vLLM integration changes, `service_manager.py` for process lifecycle management, and archetype derivation logic for data consistency between TypeScript and Python systems

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant CLI as "Babylon CLI"
    participant Engine as "GameWorld Engine" 
    participant TrajectoryEngine as "TrajectoryMarketEngine"
    participant DB as "PostgreSQL"
    participant ServiceManager as "ServiceManager"
    participant AtroposAPI as "Atropos API"
    participant vLLM as "vLLM Server"
    participant BabylonEnv as "Babylon RLAIF Env"
    participant Trainer as "GRPO Trainer"
    participant WandB as "W&B (optional)"

    User->>CLI: "babylon train parallel --archetypes trader --num-agents 5 --ticks 20"
    CLI->>Engine: "Generate trajectories"
    
    loop For each agent/tick
        Engine->>TrajectoryEngine: "generateBatchDecisions()"
        TrajectoryEngine->>TrajectoryEngine: "Record trajectory with archetype"
        TrajectoryEngine->>Engine: "Return trading decisions"
    end
    
    Engine->>DB: "Save trajectories with stepsJson"
    CLI->>User: "Trajectory generation complete"

    User->>CLI: "python scripts/run_training.py --steps 100"
    CLI->>ServiceManager: "start_all()"
    ServiceManager->>AtroposAPI: "Start Atropos API server"
    ServiceManager->>vLLM: "Start vLLM inference server"
    
    CLI->>BabylonEnv: "start_environment()"
    BabylonEnv->>DB: "Load trajectories by window"
    BabylonEnv->>BabylonEnv: "Score with archetype-aware judge"
    
    CLI->>Trainer: "start_trainer()"
    Trainer->>AtroposAPI: "register()"
    
    loop Training steps
        Trainer->>AtroposAPI: "get_batch()"
        AtroposAPI->>BabylonEnv: "Request scored trajectories"
        BabylonEnv->>vLLM: "Generate completions for rollouts"
        vLLM->>BabylonEnv: "Model responses"
        BabylonEnv->>BabylonEnv: "Score using archetype composite reward"
        BabylonEnv->>AtroposAPI: "Return ScoredDataGroup"
        AtroposAPI->>Trainer: "Batch of scored trajectories"
        Trainer->>Trainer: "GRPO training step"
        Trainer->>WandB: "Log metrics (optional)"
        
        alt Every N steps
            Trainer->>Trainer: "save_checkpoint()"
            Trainer->>ServiceManager: "restart_vllm() with new weights"
        end
    end
    
    Trainer->>User: "Training complete"
```

<details><summary><h3>Context used (3)</h3></summary>

- Context from `dashboard` - CLAUDE.md ([source](https://app.greptile.com/review/custom-context?memory=d07fd731-326a-4ea1-84d3-eec6a8044b74))
- Context from `dashboard` - .cursorrules ([source](https://app.greptile.com/review/custom-context?memory=d8da6060-b92a-430f-81d2-446cd9e51763))
- Context from `dashboard` - AGENTS.md ([source](https://app.greptile.com/review/custom-context?memory=445626a4-e384-4d3b-99de-4ccf9a293a0d))
</details>

### CODE REVIEW

**Review by SYMBaiEX — DISMISSED** [review]
image

## ✅ Code Review Approved

**Local Verification Results:**
- `bun run typecheck` ✅ Passed
- `bun run lint` ✅ Passed (0 errors/warnings)
- `bun run build` ✅ Passed
- Python unit tests: **238 passed** ✅
- TypeScript unit tests: **107 passed** ✅

### Summary

Comprehensive PR that adds production-ready local training infrastructure. The review feedback has been addressed across 3 fix commits:

- ✅ `TrainingError` now inherits from `Exception`
- ✅ vLLM URL configurable via `server_configs`
- ✅ File handle leaks fixed in `service_manager.py`
- ✅ Signal handler restoration uses try/finally
- ✅ Bounded error list with `MAX_ERRORS_IN_MEMORY=200`
- ✅ Comprehensive `TESTING.md` documentation

### Test Coverage
Excellent 4-tier testing approach with 345 total tests passing locally:
- Tier 1: Unit tests (no infrastructure)
- Tier 2: JSON mode (filesystem only)
- Tier 3: DB integration (Docker PostgreSQL)
- Tier 4: Full stack (GPU training)

LGTM 🚀
